libraries/AP_CRSF/AP_CRSF_Out.cpp

```diff
@@ -0,0 +1,437 @@
+/*
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * AP_CRSF_Out.cpp - High-level driver for CRSF RC Output
+ *
+ * This class provides the high-level "application" logic for the
+ * CRSF RC Output feature. It is responsible for reading servo output values
+ * from the main SRV_Channels and telling its underlying CRSF protocol instance
+ * to send them at a user-configurable rate.
+ */
+#include "AP_CRSF_config.h"
+
+#if AP_CRSF_OUT_ENABLED
+
+#include "AP_CRSF_Out.h"
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <SRV_Channel/SRV_Channel.h>
+#include <AP_RCProtocol/AP_RCProtocol_CRSF.h>
+#include <GCS_MAVLink/GCS.h>
+
+//#define CRSF_RCOUT_DEBUG
+//#define CRSF_RCOUT_DEBUG_FRAME
+#ifdef CRSF_RCOUT_DEBUG
+# include <AP_HAL/AP_HAL.h>
+extern const AP_HAL::HAL& hal;
+static uint32_t last_update_debug_ms = 0;
+static uint32_t num_frames = 0;
+# define debug_rcout(fmt, args...) do { if (hal.console) { hal.console->printf("CRSF_OUT: " fmt "\n", ##args); } } while(0)
+#else
+# define debug_rcout(fmt, args...)
+#endif
+
+#define DEFAULT_CRSF_OUTPUT_RATE      250U // equivalent to tracer
+
+AP_CRSF_Out* AP_CRSF_Out::_singleton;
+
+extern const AP_HAL::HAL& hal;
+
+AP_CRSF_Out::AP_CRSF_Out(AP_HAL::UARTDriver* uart, uint8_t instance, AP_CRSF_OutManager& frontend) :
+    _instance_idx(instance), _uart(uart), _frontend(frontend)
+{
+    if (_singleton == nullptr) {
+        _singleton = this;
+    }
+    init(uart);
+}
+
+// Initialise the CRSF output driver
+bool AP_CRSF_Out::init(AP_HAL::UARTDriver* uart)
+{
+    if (_state != State::WAITING_FOR_PORT) {
+        return false;
+    }
+
+    _crsf_port = NEW_NOTHROW AP_RCProtocol_CRSF(AP::RC(), AP_RCProtocol_CRSF::PortMode::DIRECT_RCOUT, uart);
```

Comment:
There should be a common base class for this and AP_RCProtocol_CRSF - we shouldn't be creating an RC protocol backend here.